normalizer.go

```diff
@@ -0,0 +1,179 @@
+	// This is "less" performant than token by token processing,
+	// but it is much simpler to implement and maintain.
+	// The trade off made here is assuming normalization runs on backend
+	// where performance is not as critical as the agent.
+	normalizedSQL = groupObfuscatedValues(normalizedSQL)
+	if !n.config.KeepSQLAlias {
+		normalizedSQL = DiscardSQLAlias(normalizedSQL)
+	}
+
+	// Dedupe collected metadata
+	dedupeNormalizedInfo(normalizedInfo)
+
+	return strings.TrimSpace(normalizedSQL), normalizedInfo, nil
+}
+
+func writeNormalizedSQL(token *Token, lastToken *Token, normalizedSQL *string) {
+	if token.Type == WS || token.Type == COMMENT || token.Type == MULTILINE_COMMENT {
+		// We don't rely on the WS token to determine if we should add a whitespace
+		return
+	}
+
+	// determine if we should add a whitespace
+	writeWhitespace(lastToken, token, normalizedSQL)
+
+	// UPPER CASE SQL keywords
+	if isSQLKeyword(token) {
+		*normalizedSQL += strings.ToUpper(token.Value)
+		return
+	}
+	*normalizedSQL += token.Value
+}
+
+// groupObfuscatedValues groups consecutive obfuscated values in a SQL query into a single placeholder.
+// It replaces "(?, ?, ...)" and "[?, ?, ...]" with "( ? )" and "[ ? ]", respectively.
+// Returns the modified SQL query as a string.
+func groupObfuscatedValues(input string) string {
+	groupable_regex := regexp.MustCompile(`(\()\s*\?(?:\s*,\s*\?\s*)*\s*(\))|(\[)\s*\?(?:\s*,\s*\?\s*)*\s*(\])`)
+	grouped := groupable_regex.ReplaceAllStringFunc(input, func(match string) string {
+		if match[0] == '(' {
+			return ArrayPlaceholder
+		}
+		return BracketPlaceholder
+	})
+	return grouped
+}
+
+// DiscardSQLAlias removes any SQL alias from the input string and returns the modified string.
+// It uses a regular expression to match the alias pattern and replace it with an empty string.
+// The function is case-insensitive and matches the pattern "AS <alias_name>".
+// The input string is not modified in place.
+func DiscardSQLAlias(input string) string {
+	return regexp.MustCompile(`(?i)\s+AS\s+[\w?]+`).ReplaceAllString(input, "")
+}
+
+func dedupeCollectedMetadata(metadata []string) []string {
+	// Dedupe collected metadata
+	// e.g. [SELECT, JOIN, SELECT, JOIN] -> [SELECT, JOIN]
+	//
+	var dedupedMetadata = []string{}
+	var metadataSeen = make(map[string]bool)
+	for _, m := range metadata {
+		if _, seen := metadataSeen[m]; !seen {
+			metadataSeen[m] = true
+			dedupedMetadata = append(dedupedMetadata, m)
+		}
+	}
+	return dedupedMetadata
+}
+
+func dedupeNormalizedInfo(info *NormalizedInfo) {
+	info.Tables = dedupeCollectedMetadata(info.Tables)
+	info.Comments = dedupeCollectedMetadata(info.Comments)
+	info.Commands = dedupeCollectedMetadata(info.Commands)
+}
+
+func isSQLKeyword(token *Token) bool {
+	return token.Type == IDENT && keywordsRegex.MatchString(token.Value)
+}
+
+func writeWhitespace(lastToken *Token, token *Token, normalizedSQL *string) {
```

Comment:
Maybe change this to `appendWhitespace(lastToken *Token, token *Token, statement string) (normalized string)`. I think `append` can follow the styling/api of the slice append operation and the usage would naturally be:

```
normalizedQuery = appendWhitespace(lastToken, token, normalizedQuery)
```